aws/resource_aws_globalaccelerator_listener.go

```diff
@@ -0,0 +1,270 @@
+				ValidateFunc: validation.StringInSlice([]string{
+					globalaccelerator.ProtocolTcp,
+					globalaccelerator.ProtocolUdp,
+				}, false),
+			},
+			"port_range": {
+				Type:     schema.TypeList,
+				Required: true,
+				MinItems: 1,
+				MaxItems: 10,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"from_port": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+						"to_port": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceAwsGlobalAcceleratorListenerCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).globalacceleratorconn
+
+	opts := &globalaccelerator.CreateListenerInput{
+		AcceleratorArn:   aws.String(d.Get("accelerator_arn").(string)),
+		IdempotencyToken: aws.String(resource.UniqueId()),
+		Protocol:         aws.String(d.Get("protocol").(string)),
+		PortRanges:       resourceAwsGlobalAcceleratorListenerExpandPortRanges(d.Get("port_range").([]interface{})),
+	}
+
+	if v, ok := d.GetOk("client_affinity"); ok {
+		opts.ClientAffinity = aws.String(v.(string))
+	}
+
+	log.Printf("[DEBUG] Create Global Accelerator listener: %s", opts)
+
+	resp, err := conn.CreateListener(opts)
+	if err != nil {
+		return fmt.Errorf("Error creating Global Accelerator listener: %s", err)
+	}
+
+	d.SetId(*resp.Listener.ListenerArn)
+
+	// Creating a listener triggers the accelerator to change status to InPending
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{globalaccelerator.AcceleratorStatusInProgress},
+		Target:  []string{globalaccelerator.AcceleratorStatusDeployed},
+		Refresh: resourceAwsGlobalAcceleratorAcceleratorStateRefreshFunc(conn, d.Get("accelerator_arn").(string)),
+		Timeout: d.Timeout(schema.TimeoutCreate),
+	}
+
+	log.Printf("[DEBUG] Waiting for Global Accelerator listener (%s) availability", d.Id())
+	_, err = stateConf.WaitForState()
+	if err != nil {
+		return fmt.Errorf("Error waiting for Global Accelerator listener (%s) availability: %s", d.Id(), err)
+	}
+
+	return resourceAwsGlobalAcceleratorListenerRead(d, meta)
+}
+
+func resourceAwsGlobalAcceleratorListenerRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).globalacceleratorconn
+
+	listener, err := resourceAwsGlobalAcceleratorListenerRetrieve(conn, d.Id())
+
+	if err != nil {
+		return fmt.Errorf("Error reading Global Accelerator listener: %s", err)
+	}
+
+	if listener == nil {
+		log.Printf("[WARN] Global Accelerator listener (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	acceleratorArn, err := resourceAwsGlobalAcceleratorListenerParseAcceleratorArn(d.Id())
+
+	if err != nil {
+		return err
+	}
+
+	d.Set("accelerator_arn", acceleratorArn)
+	d.Set("client_affinity", listener.ClientAffinity)
+	d.Set("protocol", listener.Protocol)
+	d.Set("port_range", resourceAwsGlobalAcceleratorListenerFlattenPortRanges(listener.PortRanges))
```

Comment:
When using `d.Set()` with aggregate types (TypeList, TypeSet, TypeMap), [we should perform error checking to prevent issues where the code is not properly able to set the Terraform state](https://www.terraform.io/docs/extend/best-practices/detecting-drift.html#error-checking-aggregate-types). e.g.

```suggestion
	if err := d.Set("port_range", resourceAwsGlobalAcceleratorListenerFlattenPortRanges(listener.PortRanges)); err != nil {
		return fmt.Errorf("error setting routing_config: %s", err)
	}
```